### PR TITLE
Msal node with state

### PR DESCRIPTION
**Pull request by mpminayo**
Added OAuth2.0 state feature.

### CODE REVIEW

File: lib/msal-node/src/client/ClientApplication.ts [hunk 1]
```diff
@@ -281,6 +284,20 @@ export abstract class ClientApplication {
         return this.tokenCache;
     }
 
+    /**
+     * @param state
+     * @param cachedState
+     */
+    validateState(state:string, cachedState:string): void {
+        if(!state) {
+            throw NodeAuthError.createHashDoesNotContainStateError();
```
Comment (jasonnutter) [inline lib/msal-node/src/client/ClientApplication.ts]: MSAL Node doesn't work with hashes, so the wording here should be changed. Maybe request does not contain state?

File: lib/msal-node/src/error/NodeAuthError.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+/**
+ * NodeAuthErrorMessage class containing string constants used by error codes and messages.
+ */
+export const NodeAuthErrorMessage = {
+    hashEmptyError: {
+        code: "hash_empty_error",
+        desc: "Hash value cannot be processed because it is empty. Please verify that your redirectUri is not clearing the hash."
+    },
+    StateNotFoundError: {
```
Comment (sameerag) [inline lib/msal-node/src/error/NodeAuthError.ts]:
```suggestion
    stateNotFoundError: {
```

File: lib/msal-node/src/error/NodeAuthError.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+ * Licensed under the MIT License.
+ */
+
+import { AuthError } from "@azure/msal-common";
+
+/**
+ * NodeAuthErrorMessage class containing string constants used by error codes and messages.
+ */
+export const NodeAuthErrorMessage = {
+    hashEmptyError: {
```
Comment (sameerag) [inline lib/msal-node/src/error/NodeAuthError.ts]: We never process `hash` in `msal-node`. This is done by the `user` passing us the auth code. I am curious why we need this error in `msal-node`

File: samples/msal-node-samples/auth-code/index.js [hunk 4]
```diff
@@ -105,8 +120,21 @@ const getTokenAuthCode = function (scenarioConfig, clientApplication, port) {
...
-        const authCodeResponse = { nonce: req.session.nonce, code: req.query.code }
+        const state = req.query.state
+        const cachedState = req.session.state
+
+        /**
+         * MSAL Node provides the ClientApplication.validateSate API, which validates the state string received when 
+         * authorization code is returned.
+         *
+         */
+        clientApplication.validateState(state, cachedState)
```
Comment (sameerag) [inline samples/msal-node-samples/auth-code/index.js]:
```suggestion
        clientApplication.validateState(state, cachedState);
```

File: samples/msal-node-samples/auth-code/index.js [hunk 5]
```diff
@@ -105,8 +120,21 @@ const getTokenAuthCode = function (scenarioConfig, clientApplication, port) {
...
+        const tokenRequest = { ...requestConfig.tokenRequest, code: req.query.code, state:req.query.state };
+        const authCodeResponse = { 
+            nonce: req.session.nonce, 
+            code: req.query.code 
+        }
```
Comment (sameerag) [inline samples/msal-node-samples/auth-code/index.js]:
```suggestion
        };
```

**Review by sameerag — CHANGES_REQUESTED** [review]

File: samples/msal-node-samples/auth-code/index.js [hunk 6]
```diff
@@ -58,7 +58,21 @@ const getTokenAuthCode = function (scenarioConfig, clientApplication, port) {
...
+             * The generated state is then cached and passed as part of authCodeUrlParameters during authentication request.
+             * 
+             * For more information about state,
+             * visit https://datatracker.ietf.org/doc/html/rfc6819#section-3.6
+             */
+            if(req.query.state) {
+                authCodeUrlParameters.state = req.query.state;
+            } else {
+                authCodeUrlParameters.state = cryptoProvider.createNewGuid()
+            }
```
Comment (sameerag) [inline samples/msal-node-samples/auth-code/index.js]:
```suggestion
authCodeUrlParameters.state = req.query.state ? req,query.state : cryptoProvider.createNewGuid();
```

File: samples/msal-node-samples/auth-code/index.js [hunk 7]
```diff
@@ -105,8 +120,21 @@ const getTokenAuthCode = function (scenarioConfig, clientApplication, port) {
...
-        const authCodeResponse = { nonce: req.session.nonce, code: req.query.code }
+        const state = req.query.state
+        const cachedState = req.session.state
+
+        /**
+         * MSAL Node provides the ClientApplication.validateSate API, which validates the state string received when 
+         * authorization code is returned.
+         *
+         */
+        clientApplication.validateState(state, cachedState);
```
Comment (sameerag) [inline samples/msal-node-samples/auth-code/index.js]:
```suggestion
        clientApplication.validateState(req.query.state, req.session.state);
```
Can get rid of the const declarations above

**Review by sameerag — APPROVED** [review]
Added a few more suggestions. Approving with recommendation to make those changes.

File: samples/msal-node-samples/auth-code/index.js (hunk 6 above)
Comment (jasonnutter) [reply samples/msal-node-samples/auth-code/index.js]: Can also be: `authCodeUrlParameters.state = req.query.state || cryptoProvider.createNewGuid();`

File: change/@azure-msal-react-5a5d4ecb-9034-460b-8efc-8af3b2c2253b.json [hunk 8]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "minor",
+  "comment": "Removed hash error from msal node, fixed missing commas",
+  "packageName": "@azure/msal-react",
```
Comment (tnorling) [inline change/@azure-msal-react-5a5d4ecb-9034-460b-8efc-8af3b2c2253b.json]: Please remove this as there are no msal-react changes

File: change/@azure-msal-node-892ff60a-4f9d-47e4-8217-04e90f1381e4.json [hunk 9]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "minor",
+  "comment": "Fixing merge conflict",
```
Comment (tnorling) [inline change/@azure-msal-node-892ff60a-4f9d-47e4-8217-04e90f1381e4.json]: Please give this a more descriptive comment as this will be published in the changelog (and include the PR #)

File: lib/msal-node/src/client/ClientApplication.ts [hunk 10]
```diff
@@ -281,6 +284,20 @@ export abstract class ClientApplication {
         return this.tokenCache;
     }
 
+    /**
+     * @param state
+     * @param cachedState
+     */
+    validateState(state:string, cachedState:string): void {
```
Comment (tnorling) [inline lib/msal-node/src/client/ClientApplication.ts]: Is this being called anywhere?

File: lib/msal-node/src/error/NodeAuthError.ts [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+/**
+ * NodeAuthErrorMessage class containing string constants used by error codes and messages.
+ */
+export const NodeAuthErrorMessage = {
+    stateNotFoundError: {
+        code: "state_not_found",
+        desc: "State not found. Please verify that the request originated from msal."
+    },
+   
+    unableToParseStateError: {
```
Comment (tnorling) [inline lib/msal-node/src/error/NodeAuthError.ts]: Is this being used anywhere?

File: package-lock.json [hunk 12]
```diff
@@ -1,15779 +1,7 @@
 {
   "name": "microsoft-authentication-libraries-for-js",
-  "lockfileVersion": 2,
```
Comment (tnorling) [inline package-lock.json]: Please discard these changes and upgrade your local node version to 16 to avoid lock file churn

**Review by tnorling — CHANGES_REQUESTED** [review]

File: samples/msal-node-samples/auth-code/index.js (hunk 6 above)
Comment (Robbie-Microsoft) [reply samples/msal-node-samples/auth-code/index.js]: +1 for Jason's suggestion

**Comment by peterzenz** [comment]
@mpminayo Any updates here?

File: lib/msal-node/src/error/NodeAuthError.ts [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+        code: "unable_to_parse_state",
```
Comment (bmahall) [inline lib/msal-node/src/error/NodeAuthError.ts]: Any relevant tests would be helpful. Thanks

File: lib/msal-node/test/client/PublicClientApplication.spec.ts [hunk 14]
```diff
@@ -412,4 +411,33 @@ describe('PublicClientApplication', () => {
...
+                "ed09b151-1b68-4c2c-8e95-d8dce9882dba",
+                "ed09b151-1b68-4c2c-8e95-y8dcfffffggh"
+            );
+        }).toThrow(
+            "state_mismatch: State mismatch error. Please check your network. Continued requests may cause cache overflow"
+        );
+    })
 });
+
+
```
Comment (jo-arroyo) [inline lib/msal-node/test/client/PublicClientApplication.spec.ts:521]: Please remove extra lines.

File: lib/msal-node/src/client/ClientApplication.ts [hunk 15]
```diff
@@ -281,6 +284,20 @@ export abstract class ClientApplication {
         return this.tokenCache;
     }
 
+    /**
+     * @param state
```
Comment (jo-arroyo) [inline lib/msal-node/src/client/ClientApplication.ts:299]: Please add a description here for typedoc purposes, including how this API can be used.

File: change/@azure-msal-node-892ff60a-4f9d-47e4-8217-04e90f1381e4.json (hunk 9 above)
Comment (jo-arroyo) [reply change/@azure-msal-node-892ff60a-4f9d-47e4-8217-04e90f1381e4.json]: +1

File: lib/msal-node/src/client/ClientApplication.ts (hunk 10 above)
Comment (jo-arroyo) [inline lib/msal-node/src/client/ClientApplication.ts]:
```suggestion
    validateState(state: string, cachedState: string): void {
```
Nit

File: samples/msal-node-samples/auth-code/index.js [hunk 16]
```diff
@@ -109,8 +116,18 @@ const getTokenAuthCode = function (scenarioConfig, clientApplication, port) {
     });
 
     app.get("/redirect", (req, res) => {
-        const tokenRequest = { ...requestConfig.tokenRequest, code: req.query.code };
-        const authCodeResponse = { nonce: req.session.nonce, code: req.query.code }
+        /**
+         * MSAL Node provides the ClientApplication.validateSate API, which validates the state string received when 
+         * authorization code is returned.
+         *
```
Comment (jo-arroyo) [inline samples/msal-node-samples/auth-code/index.js]: Extra lines can be removed

File: lib/msal-node/src/error/NodeAuthError.ts (hunk 11 above)
Comment (jo-arroyo) [reply lib/msal-node/src/error/NodeAuthError.ts]: +1

File: lib/msal-node/src/client/ClientApplication.ts (hunk 10 above)
Comment (mpminayo) [reply lib/msal-node/src/client/ClientApplication.ts]: In auth-code sample.

File: lib/msal-node/src/error/NodeAuthError.ts (hunk 11 above)
Comment (mpminayo) [reply lib/msal-node/src/error/NodeAuthError.ts]: In validateState method in ClientApplication.ts

File: lib/msal-node/src/client/ClientApplication.ts (hunk 10 above)
Comment (mpminayo) [reply lib/msal-node/src/client/ClientApplication.ts]: Yes, it's added in the auth-code sample.

File: lib/msal-node/src/error/NodeAuthError.ts (hunk 11 above)
Comment (tnorling) [reply lib/msal-node/src/error/NodeAuthError.ts]: I see the `stateNotFoundError` in ClientApplication, but not this one?

File: lib/msal-node/src/client/ClientApplication.ts (hunk 10 above)
Comment (tnorling) [reply lib/msal-node/src/client/ClientApplication.ts]: I see. I'm not sure that this should be a public API. We should handle this for people under the hood.

File: lib/msal-node/src/error/NodeAuthError.ts (hunk 11 above)
Comment (mpminayo) [reply lib/msal-node/src/error/NodeAuthError.ts]: I am actually not using unableToParseStateError, instead, I am using StateMismatchError from msal common.

File: lib/msal-node/src/client/ClientApplication.ts (hunk 10 above)
Comment (mpminayo) [reply lib/msal-node/src/client/ClientApplication.ts]: I have moved this to be called from msal-node.

File: change/@azure-msal-node-892ff60a-4f9d-47e4-8217-04e90f1381e4.json [hunk 17]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "minor",
+  "comment": "Fixing conflicts in msal-node/src/client/ClientApplication.ts after merging dev into msal-node-with-state pr# ccd14d2ff",
```
Comment (tnorling) [inline change/@azure-msal-node-892ff60a-4f9d-47e4-8217-04e90f1381e4.json]:
```suggestion
  "comment": "Add support for the state parameter #4723",
```

File: lib/msal-node/src/client/ClientApplication.ts [hunk 18]
```diff
@@ -132,11 +134,16 @@ export abstract class ClientApplication {
      */
     async acquireTokenByCode(request: AuthorizationCodeRequest, authCodePayLoad?: AuthorizationCodePayload): Promise<AuthenticationResult | null> {
         this.logger.info("acquireTokenByCode called", request.correlationId);
+        if (request.state && authCodePayLoad){
+            this.validateState(request.state, authCodePayLoad.state || "");
+            authCodePayLoad.state = undefined;
```
Comment (tnorling) [inline lib/msal-node/src/client/ClientApplication.ts]: We should not mutate variables, why is this needed?

File: lib/msal-node/src/client/ClientApplication.ts [hunk 19]
```diff
@@ -132,11 +134,16 @@ export abstract class ClientApplication {
      */
     async acquireTokenByCode(request: AuthorizationCodeRequest, authCodePayLoad?: AuthorizationCodePayload): Promise<AuthenticationResult | null> {
         this.logger.info("acquireTokenByCode called", request.correlationId);
+        if (request.state && authCodePayLoad){
```
Comment (tnorling) [inline lib/msal-node/src/client/ClientApplication.ts:137]: Can we add a logging here?

File: lib/msal-node/src/error/NodeAuthError.ts (hunk 11 above)
Comment (tnorling) [reply lib/msal-node/src/error/NodeAuthError.ts]: If this isn't being used anywhere, please remove.

File: lib/msal-node/src/client/ClientApplication.ts (hunk 18 above)
Comment (mpminayo) [reply lib/msal-node/src/client/ClientApplication.ts]: When state is added to authCodePaylLoad, msal common's ProtocolUtils' parseRequestState method parses the library state and state sent by the user. In msal node we are not using library state, so to avoid ClientAuthError.createInvalidStateError, I am setting the state to undefined.

Comment (tnorling) [reply lib/msal-node/src/client/ClientApplication.ts]: If you need to do this you should create a new variable (using the spread operator) rather than editing the one passed in

**Review by bktowett — APPROVED** [review]

**Review by bgavrilMS — APPROVED** [review]

**Review by hectormmg — APPROVED** [review]

**Review by bmahall — APPROVED** [review]

File: change/@azure-msal-node-892ff60a-4f9d-47e4-8217-04e90f1381e4.json (hunk 17 above)
Comment (tnorling) [inline change/@azure-msal-node-892ff60a-4f9d-47e4-8217-04e90f1381e4.json]:
```suggestion
  "comment": "Add support for the state parameter #4723",
```

File: lib/msal-node/src/client/ClientApplication.ts [hunk 20]
```diff
@@ -132,11 +134,18 @@ export abstract class ClientApplication {
      */
     async acquireTokenByCode(request: AuthorizationCodeRequest, authCodePayLoad?: AuthorizationCodePayload): Promise<AuthenticationResult> {
         this.logger.info("acquireTokenByCode called", request.correlationId);
+        if (request.state && authCodePayLoad){
+            this.logger.info("validating state");
```
Comment (tnorling) [inline lib/msal-node/src/client/ClientApplication.ts]:
```suggestion
            this.logger.info("acquireTokenByCode - validating state");
```

File: lib/msal-node/src/client/ClientApplication.ts [hunk 21]
```diff
@@ -281,6 +290,25 @@ export abstract class ClientApplication {
...
+    /**
+     * Validates OIDC state by comparing the user cached state with the state received from the server.
+     * 
+     * This API is provided for scenarios where you would use OAuth2.0 state parameter to mitigate against
+     * CSRF attacks.
+     * For more information about state, visit https://datatracker.ietf.org/doc/html/rfc6819#section-3.6.
+     * @param state
+     * @param cachedState
+     */
+    validateState(state: string, cachedState: string): void {
```
Comment (tnorling) [inline lib/msal-node/src/client/ClientApplication.ts]:
```suggestion
    protected validateState(state: string, cachedState: string): void {
```

**Review by tnorling — APPROVED** [review]
